music21/expressions.py

```diff
@@ -1266,30 +1619,39 @@ def realize(
                 raise ExpressionException('The note is not long enough for a nachschlag')
             useQL = srcObj.duration.quarterLength / 4
 
-        transposeInterval = self.getSize(srcObj.pitch)
+        currentKeySig = keySig
+        if currentKeySig is None:
+            currentKeySig = srcObj.getContextByClass(key.KeySignature)
+            if currentKeySig is None:
+                currentKeySig = key.KeySignature(0)
+
+        transposeInterval = self.getSize(srcObj, keySig=currentKeySig)
         transposeIntervalReverse = transposeInterval.reverse()
+        isTransposed: bool = transposeInterval != interval.Interval('P1')
 
         numberOfTrillNotes = int(srcObj.duration.quarterLength / useQL)
         if self.nachschlag:
             numberOfTrillNotes -= 2
 
-        trillNotes: list[note.Note] = []
+        trillNotes: list[note.Note | note.Unpitched] = []
         for unused_counter in range(int(numberOfTrillNotes / 2)):
             self.fillListOfRealizedNotes(srcObj, trillNotes, transposeInterval, useQL=useQL)
 
-        currentKeySig = None
-        setAccidentalFromKeySig = self._setAccidentalFromKeySig
-        if setAccidentalFromKeySig:
-            currentKeySig = srcObj.getContextByClass(key.KeySignature)
-            if currentKeySig is None:
-                currentKeySig = key.KeySignature(0)
-
-            for n in trillNotes:
-                if n.pitch.nameWithOctave != srcObj.pitch.nameWithOctave:
-                    # do not correct original note, no matter what.
-                    if n.pitch.accidental is None:
-                        # only correct if there isn't already an accidental (from accid)
-                        n.pitch.accidental = currentKeySig.accidentalByStep(n.step)
+        if isTransposed:
+            setAccidentalFromKeySig = self._setAccidentalFromKeySig
+            if setAccidentalFromKeySig:
+                for n in trillNotes:
+                    if t.TYPE_CHECKING:
+                        # for some reason, this assertion doesn't convince mypy that n
+                        # has a pitch attribute, so we have to "type: ignore" below.
+                        assert isinstance(n, note.Note)
+                    if n.pitch.nameWithOctave != srcObj.pitch.nameWithOctave:  # type: ignore
```

Comment:
mypy should be convinced -- the problem is that srcObj might be a Unpitched object

Reply:
It's not n it's srcObj I'm guessing.  You never asserted that IT was a note.

Reply:
Oh. Duh. 🤦🏻‍♂️

Reply:
Done.